web/router.go

```diff
@@ -38,10 +43,64 @@ func SetupRoutes(router *gin.Engine) {
 	router.Use(middlewares.ParseHost())
 	router.Use(middlewares.ServeApp(apps.Serve))
 	router.Use(middlewares.ErrorHandler())
+	router.Use(corsMiddleware("/apps", "/data", "/files"))
+
 	auth.Routes(router)
 	apps.Routes(router.Group("/apps", middlewares.NeedInstance()))
 	data.Routes(router.Group("/data", middlewares.NeedInstance()))
 	files.Routes(router.Group("/files", middlewares.NeedInstance()))
 	status.Routes(router.Group("/status"))
 	version.Routes(router.Group("/version"))
 }
+
+func corsMiddleware(routes ...string) gin.HandlerFunc {
+	normalHeaders := make(http.Header)
+	normalHeaders.Set("Access-Control-Allow-Credentials", "true")
+
+	preflightHeaders := make(http.Header)
+	preflightHeaders.Set("Access-Control-Allow-Credentials", "true")
+	preflightHeaders.Set("Access-Control-Allow-Methods", "*")
+	preflightHeaders.Set("Access-Control-Max-Age", strconv.FormatInt(int64(12*time.Hour/time.Second), 10))
+
+	return func(c *gin.Context) {
+		origin := c.Request.Header.Get("Origin")
+		if origin == "" {
+			return
+		}
```

Comment:
TODO for later: check that the `Origin` header matches a subdomain of the instance or a registered domain for an OAuth2 app/device.